trezor/trezor-suite
# metadata: encryption version migration

Based on #8646  which is a PR that introduces changes to metadata data model in reducers. That PR is basically ready to be merged but as it does not bring any new features I am hesitant to do it. Also I dont want to bloat it with more changes so I decided to build on it instead. I am also stealing some ideas from https://github.com/trezor/trezor-suite/pull/8567

### Commits
- 1st commit - is basically fixup for #8646 I realized that `device.metadata.key` should also be versioned as it determines `account.metadata.aesKey` and `account.metadata.fileName`. This fixup is incomplete, there are also some changes in the second commit that should go here.
- 2nd commit -implements metadata migration and resolves #6304 
- 3rd commit does not exist yet. I would definitely add at least e2e test from here https://github.com/trezor/trezor-suite/pull/8567 and whatevere else might be useful

### Todos: 
- [x] couple of todos and areas for improvements

### How does it work
1.  user tries to add a label, initation process kicks in as usual
2. once data provider is connected and metadata keys (for the latest encryption version) are derived we fetch list of all files from data provider
3. if all labelable entities (currently discovered) have associated file in what we fetchted in 2 no further action is needed
4. else we separate labelable entities that don't have their counterpart in dataprovider and derive legacy keys for them
5. then we fetch legacy files from data provider
6. we decrypt legacy files using legacy keys, load them into application memory and encrypt them using new keys and upload them to data provider. 
7. if there was no legacy file match in 5, we create dummy files and upload them as well. this ensures that next time metadata is initiated, we are likely to be able to stop at point 3 (unless user created a new account in the meantime somewhere else)

### Testing tips
1. create an empty account on dropbox
8. go to https://suite.trezor.io/web
9. create some labels
10. open build of this branch 
11. enable labeling
12. labels should be there
13. forget device, enable labeling again. you should not be prompted on device. labels should still be there

### Drawbacks
- we can not migrate all labels at once because we simply can't be sure we know them. Labels are bound to 'labelable entities' which are now accounts and wallets. So whenever user adds a labelable entity, we must run "metadata discovery" including migration if needed. So it is very well possible that user might migrate his data on day, they might be using suite without "enable labeling prompt" for years, and then, after adding some long unused passphrase or account, user will need to confirm labeling on device again.
- possibility of mismatch between versions of suite. after this is release user opens updated version of suite, migrates data, then opens outdated version of suite and works with labeling. Under current implementation labels added in the outdated version will be lost.
